src/flow_graph.py

```diff
@@ -1096,15 +1274,163 @@ def reset_block_info(self) -> None:
             node.block.block_info = None
 
 
+def phi_reg_sources(node: Node, reg: Access, imdom_srcs: RefSet) -> RefSet:
+    """
+    Return the RefSet of all of the places `reg` is assigned to, if it is a valid phi.
+    Otherwise, return RefSet.invalid().
+    """
+    if node.immediate_dominator is None:
```

Comment:
Should be possible to assert that this is not None, given the caller. (If it *is* None I think the right answer might be to include references to arguments etc. rather than return invalid? but we don't have to think about that)

Reply:
Completely unrelated, but I just realized we're assuming everywhere that the entry node has no parents/dominators, but I don't think we ever make sure that that's the case? (unless I missed some check in my cursory review) We might want to add an assert for that. (Or more ambitiously, inject an empty initial block if the first block has parents.)

Reply:
Yeah you're right -- the existing code assumes the entry node has no parents. We should have to set up phis for the entry node if it has parents. 

I don't think the entry node can ever have an explicit dominator though? Its caller always acts as an implicit parent/immediate dominator?